subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/bytebuffer/ByteBufferPersistedData.java

```diff
@@ -0,0 +1,236 @@
+        }
+        throw new ClassCastException("Data is not a number");
+    }
+
+    @Override
+    public int getAsInteger() {
+        if (!isNumber()) {
+            throw new ClassCastException("It is not number");
+        }
+        resetPosition();
+        switch (type) {
+            case INTEGER:
+                return getData().getInt();
+            case FLOAT:
+                return (int) getData().getFloat();
+            case DOUBLE:
+                return (int) getData().getDouble();
+            case LONG:
+                return (int) getData().getLong();
+        }
+        throw new ClassCastException("Data is not a number");
+    }
+
+    @Override
+    public long getAsLong() {
+        if (!isNumber()) {
+            throw new ClassCastException("It is not number");
+        }
+        resetPosition();
+        switch (type) {
+            case INTEGER:
+                return getData().getInt();
+            case FLOAT:
+                return (long) getData().getFloat();
+            case DOUBLE:
+                return (long) getData().getDouble();
+            case LONG:
+                return getData().getLong();
+        }
+        throw new ClassCastException("Data is not a number");
+    }
+
+
+    @Override
+    public boolean getAsBoolean() {
+        if (!isBoolean()) {
+            throw new ClassCastException("It is not boolean");
+        }
+        resetPosition();
+        return byteBuffer.get() != 0; // Don't Packed booleans
+    }
+
+    @Override
+    public byte[] getAsBytes() {
+        if (!isBytes()) {
+            throw new DeserializationException("it is not bytes or bytebuffer");
+        }
+        resetPosition();
+        int size = byteBuffer.getInt();
+        byte[] bytes = new byte[size];
+        byteBuffer.get(bytes);
+        return bytes;
+    }
+
+    @Override
+    public ByteBuffer getAsByteBuffer() {
+        if (!isBytes()) {
+            throw new DeserializationException("it is not bytes or bytebuffer");
+        }
+        resetPosition();
+        return ByteBuffer.wrap(getAsBytes());
+    }
+
+    @Override
+    public PersistedDataArray getAsArray() {
+        if (!isArray()) {
+            throw new IllegalStateException("it is not array");
+        }
+        byteBuffer.position(position);
+        return new ByteBufferPersistedDataArray(byteBuffer);
+    }
+
+    @Override
+    public PersistedDataMap getAsValueMap() {
+        if (!isValueMap()) {
+            throw new IllegalStateException("it is not valuemap");
+        }
+        byteBuffer.position(position);
+        return new ByteBufferPersistedDataMap(byteBuffer);
+    }
+
+    @Override
+    public boolean isString() {
+        return type == BBType.STRING;
+    }
+
+    @Override
+    public boolean isNumber() {
+        return type == BBType.FLOAT
+                || type == BBType.DOUBLE
+                || type == BBType.INTEGER
+                || type == BBType.LONG;
```

Comment:
you can define all this encoding information in BBType

Reply:
Yeah. but not directly.
I cannot use enum values in same enum 'constructors'

Another variant, i can provide something like capabilities.. but it is will mess `NULL(0,false,false,false,false,false,true)`

or switches... switches... switches in methods..